backend/src/nodes/image_util_nodes.py

```diff
@@ -12,110 +12,112 @@
-        ov1: np.ndarray,
-        op1: float,
-        ov2: Union[np.ndarray, None],
-        op2: float,
+        ov: np.ndarray,
+        op: float,
+        blend_mode: int,
     ) -> np.ndarray:
-        """Overlay transparent images on base image"""
+        """Blend images together"""
 
         # Convert to 0.0-1.0 range
         opbase /= 100
-        op1 /= 100
-        op2 /= 100
+        op /= 100
 
         imgs = []
-        max_h, max_w = 0, 0
-        for img in base, ov1, ov2:
-            if img is not None:
-                h, w, c = get_h_w_c(img)
-                max_h = max(h, max_h)
-                max_w = max(w, max_w)
+        max_h = max_w = 0
+        for img in (base, ov):
+            h, w, c = get_h_w_c(img)
+            max_h = max(h, max_h)
+            max_w = max(w, max_w)
 
-                # All inputs must be BGRA for alpha compositing to work
-                if c == 1:
-                    img = cv2.cvtColor(img, cv2.COLOR_GRAY2BGRA)
-                elif c == 3:
-                    img = np.dstack((img, np.ones((h, w), np.float32)))
-                elif c != 4:  # Explode if there are not 1, 3, or 4 channels
-                    logger.error(f"Number of channels ({c}) unexpected")
+            # All inputs must be BGRA for alpha compositing to work
+            if c == 1:
+                img = cv2.cvtColor(img, cv2.COLOR_GRAY2BGRA)
+            elif c == 3:
+                img = cv2.cvtColor(img, cv2.COLOR_BGR2BGRA)
+            elif c != 4:  # Explode if there are not 1, 3, or 4 channels
+                logger.error(f"Number of channels ({c}) unexpected")
 
-                imgs.append(img)
-        assert (
-            base.shape[0] >= max_h and base.shape[1] >= max_w
-        ), "Base must be largest image."
+            imgs.append(img)
 
         imgout = imgs[0]
-        imgs = imgs[1:]
+        img = imgs[1]
+
+        # Pad base image with transparency if necessary to match size with overlay
+        h, w, _ = get_h_w_c(imgout)
+        tp = bm = lt = rt = 0
+        if h < max_h:
+            tp = (max_h - h) // 2
+            bm = max_h - h - tp
+        if w < max_w:
+            lt = (max_w - w) // 2
+            rt = max_w - w - lt
+        imgout = cv2.copyMakeBorder(
+            imgout, tp, bm, lt, rt, cv2.BORDER_CONSTANT, value=0
+        )
 
+        # Center overlay
+        h, w, _ = get_h_w_c(img)
         center_x = imgout.shape[1] // 2
         center_y = imgout.shape[0] // 2
+        x_offset = center_x - (w // 2)
+        y_offset = center_y - (h // 2)
 
-        # Apply opacity to base, then overlay A and B sequentially at corresponding opacities
+        # Apply opacities to images, then blend overlay
         imgout[:, :, 3] *= opbase
-        for img, op in zip(imgs, (op1, op2)):
-            h, w, _ = get_h_w_c(img)
-
-            # Center overlay
-            x_offset = center_x - (w // 2)
-            y_offset = center_y - (h // 2)
+        img[:, :, 3] *= op
```

Comment:
Are those the original image objects? If so, then you're modifying cached images, which is a bug.

Reply:
They're the objects defined at lines 86-87:
```
imgout = imgs[0]
img = imgs[1]
```

Reply:
In pretty much any OOP language, objects are passed by reference (technically by value of their pointer, but that's close to by reference). This means you can do stuff like this:

```py
a = SomeClass()
a.foo = 9
b = a
a.foo = 10
print(a.foo) # 10
print(b.foo) # 10
```

Since you only passed around the reference of the input image, I'm worried that you are modifying the cached image directly.

@joeyballentine Asking a python expert, is that what's happening here or is there some weird python magic I am unaware of?

Reply:
So a partial answer to this is that lists are mutable in python, so yes, that does modify the input images in the case of RGBA input.

So if I did `imgout[0 ,0, 0] = 3`, then that would alter base. If I then did `imgout = np.clip(imgout, 0, 1)`, that would not alter base, and it would still have that 3. The return value is a new object, so imgout now refers to the returned object, not base.

I'd be interested to hear more about what the bug is, as to my understanding this is considered normal behavior in python (and I'm pretty sure occurs many places in chaiNNer aside from this node). I don't know what issue it would cause to modify the object passed in vs. creating a new object (which would require either functions that return a new value or a deep copy), and the nodes always return a new object.

Reply:
It's all academic anyway, factoring out the BGRA conversion means that it will be returning new objects anyway, so whether or not this is an issue, it will not be an issue.

Reply:
Hmm, actually, returning itself may not create a new object like I thought. I still am not sure what bug this would create though.

Reply:
The bug is not that lists (numpy arrays in this case) are mutable. That's by design. The problem is that the executor caches images. So an image passed to `run` might also be passed to other nodes afterward. So if you modify the image directly, you also modify that image for all nodes that get executed afterward. 

The executor is essentially doing this:
```py
img = some_node.run(...)
res1 = blend_node.run(img, ...)
res2 = some_other_node.run(img, ...)
```

So modifying any of the input objects is a big no no.

Reply:
Okay, if you check the latest commit, base is not modified (ov is once potentially, but it deadends in this node and does not move on to anywhere). If base is not BGRA, cvtColor returns a new object. Even if base is BGRA, when copyMakeBorder is used, it creates a new object that is assigned to imgout, and that is the object that is modified, so imgout never refers to base when it is modified.

Reply:
Gotcha. In that case, I need to make sure ov is never modified either. Didn't realize the executor handled it like that.

Reply:
The BGRA conversion function will always return a deep copy now, so neither cached image will ever be modified.